src/util/string.h

```diff
@@ -83,8 +83,8 @@ std::vector<T> Split(const Span<const char>& sp, char sep)
 
 [[nodiscard]] inline std::string_view RemovePrefixView(std::string_view str, std::string_view prefix)
 {
-    if (str.substr(0, prefix.size()) == prefix) {
-        return str.substr(prefix.size());
+    if (str.starts_with(prefix)) {
+        str.remove_prefix(prefix.size());
```

Comment:
* `std::string_view::remove_prefix()` itself invites mutation.
* The names `TrimStringView()` and `RemovePrefixView()` imply mutation.
* As I said in https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1686494537, I could only find one other case of `const` `string_view` parameters in the entire codebase. I'm sorry if you are used to another way. It may well be a better way, but I'd rather not change more pre-existing things like `TrimStringView()` and `RemovePrefixView()` taking non-const `string_view` copies.

I do not think introducing test cases to prove that the C++ compiler is properly copying a given parameter provides much value.

Reply:
> std::string_view::remove_prefix() itself invites mutation.

no, we din't mutate before, we returned a new view. We're also doing that now, but the copy is done implicitly in the parameters. The original `substr` way made more sense - like we're doing with `TrimStringView` as well.

Reply:
The reason for using the standard C++ `remove_prefix()` function inside of the pre-existing `RemovePrefixView()` over `substr()` was covered here:
https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1679138796 
> `base_blob` now calls `RemovePrefixView()` which it did not previously do. I want to make it as efficient as possible compared to the prior `base_blob` implementation, hence this change.

To be even more clear than I was that time - please compare the implementations of [substr()](https://github.com/gcc-mirror/gcc/blob/61715e9340ab8941d40d62158fe437e9dbe3e068/libstdc%2B%2B-v3/include/std/string_view#L334) and [remove_prefix()](https://github.com/gcc-mirror/gcc/blob/61715e9340ab8941d40d62158fe437e9dbe3e068/libstdc%2B%2B-v3/include/std/string_view#L295) in GCC.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30436/files#r1686336262

> I find it quite confusing that we're sneaking in a mutation of the copy

paplorinc, maybe it would help if you could say more specifically what you find confusing about this code? At least to me the new implementation of this function seems more straightforward than the previous implementation, though I don't think there is a big difference.

I don't love the name of the `RemovePrefixView` function, because that name sounds like it actually does modify the string view like `remove_prefix` does, instead of just returning a new string view with the prefix stripped. I think a better name for this function would just be `StripPrefix`. But that's also a subjective opinion, and renaming would be beyond the scope of this PR.

Reply:
I'm not sure what you're saying, both are the same, one creates a copy via the params (unless it's inlined, I guess, that's why I find it confusing) and modifies the copied instance, the other returns a new copied instance.
I really love speed optimizations that also make the code more readable, but this does neither.

given:
```C++
static void SetHexBenchmark(benchmark::Bench& bench)
{
    uint256 hash;
    bench.run([&] {
        hash.SetHex("1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef");
        hash.SetHex("0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef");
        ankerl::nanobench::doNotOptimizeAway(hash);
    });
}
```

with 788fe9cc9ab979c5e14f544ae05dc46436892b81 we're getting `2940379.88 ops/` on average of 2 runs:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              340.85 |        2,933,826.25 |    0.6% |      0.01 | `SetHexBenchmark`

and

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              339.34 |        2,946,933.51 |    0.1% |      0.01 | `SetHexBenchmark`



with `return str.substr(prefix.size());` it's `2936915.645 ops/s` on average for 2 runs:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              340.99 |        2,932,626.59 |    0.6% |      0.01 | `SetHexBenchmark`

and

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              340.00 |        2,941,204.70 |    0.0% |      0.01 | `SetHexBenchmark`


I.e. the difference of the averages is a lot smaller than the variation of individual runs.

which is basically the same speed as `master`:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              352.95 |        2,833,265.72 |    0.3% |      0.01 | `SetHexBenchmark`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              356.66 |        2,803,757.30 |    1.4% |      0.01 | `SetHexBenchmark`

------

let's focus on making the code cleaner and more maintainable - if it's also faster, great, but let's not make it harder to reason about an maintain for an imaginary speed gain.

Reply:
> it would help if you could say more specifically what you find confusing about this code

I'll try to rephrase.
Firstly, as you've mentioned as well, the name of the method hints at mutation, but it actually returns a copy of the view.
Second, it's inline + pass by value mutation, which implies it's not a simple search&replace type inline, since the parameters aren't consts, but it also can't just mutate the original values passed as parameters, so the copy of the parameter is implicit. Making the parameters const and returning a new view directly solves most of these, i.e. you can just copy the internals of the method and paste it to its usage (which helps in theory to understand how the method will behave).
Third, this seems like a premature and untested optimization, which should likely be done in a separate PR, in case the code isn't more readable.

Reply:
> This pull request has more than a 100 comments for less lines of code changed

I did all the changes I requested during review (even provided diffs to speed it up), they're not complicated.

Reply:
> No, from reading the function signature, one knows that a copy is being made when the argument is put in as a function parameter

but it's an inline function, there is no function parameter.

Reply:
Dropped `RemovePrefixView()` modification.